hibernate-core/src/main/java/org/hibernate/boot/model/internal/PropertyBinder.java

```diff
@@ -342,6 +336,65 @@ private Property bind(Property property) {
 		return property;
 	}
 
+	private void setDeclaredIdentifier(RootClass rootClass, MappedSuperclass superclass, Property prop) {
+		if ( superclass == null ) {
+			rootClass.setDeclaredIdentifierProperty( prop );
+			return;
+		}
+		// If the type has type parameters, we have to set the declared identifier property on the rootClass
+		// to be able to retrieve it with the correct type based on type variable assignment in the subclass
+		final Class<?> type = buildingContext.getBootstrapContext().getReflectionManager().toClass( declaringClass );
+		if ( type.getTypeParameters().length == 0 ) {
+			rootClass.setDeclaredIdentifierProperty( prop );
```

Comment:
I think this should be
```suggestion
			superclass.setDeclaredIdentifierProperty( prop );
```